lib/msal-node/src/cache/CacheManager.ts

```diff
@@ -0,0 +1,221 @@
+            let cache = Serializer.serializeAllCache(this.storage.getCache());
+
+            const getMergedState = (stateFromDisk: string) => {
+
+                if (!StringUtils.isEmpty(stateFromDisk)) {
+                    this.cacheSnapshot = stateFromDisk;
+                    cache = this.mergeState(JSON.parse(stateFromDisk), cache);
+                }
+
+                return JSON.stringify(cache);
+            };
+
+            await this.persistence.writeToStorage(getMergedState);
+            this.hasChanged = false;
+        } else {
+            throw ClientAuthError.createCachePluginError();
+        }
+    }
+
+    /**
+     * Calls ICachePlugin.readFromStorage and deserializes JSON to in-memory cache.
+     * ICachePlugin must be set on ClientApplication.
+     */
+    async readFromPersistence(): Promise<void> {
+        if (this.persistence) {
+            this.cacheSnapshot = await this.persistence.readFromStorage();
+
+            if (!StringUtils.isEmpty(this.cacheSnapshot)) {
+                const cache = this.overlayDefaults(JSON.parse(this.cacheSnapshot));
+                const deserializedCache = Deserializer.deserializeAllCache(cache);
+                this.storage.setCache(deserializedCache);
+            }
+        } else {
+            throw ClientAuthError.createCachePluginError();
+        }
+    }
+
+    /**
+     * Called when the cache has changed state.
+     */
+    private handleChangeEvent() {
+        this.hasChanged = true;
+    }
+
+    /**
+     * Merge in memory cache with the cache snapshot.
+     * @param oldState
+     * @param currentState
+     */
+    private mergeState(oldState: JsonCache, currentState: JsonCache): JsonCache {
+        let stateAfterRemoval = this.mergeRemovals(oldState, currentState);
+        return this.mergeUpdates(stateAfterRemoval, currentState);
+    }
+
+    /**
+     * Deep update of oldState based on newState values
+     * @param oldState
+     * @param newState
+     */
+    private mergeUpdates(oldState: any, newState: any): JsonCache {
+        Object.keys(newState).forEach((newKey) => {
+            let newValue = newState[newKey];
+
+            // if oldState does not contain value but newValue does, add it
+            if (!oldState.hasOwnProperty(newKey)) {
+                if (newValue != null) {
+                    oldState[newKey] = newValue;
+                }
+            } else {
+                // both oldState and newState contain the key, do deep update
+                let newValueNotNull = newValue !== null;
+                let newValueIsObject = typeof newValue === 'object';
+                let newValueIsNotArray = !Array.isArray(newValue);
+
+                if (newValueNotNull && newValueIsObject && newValueIsNotArray) {
+                    this.mergeUpdates(oldState[newKey], newValue);
+                } else {
+                    oldState[newKey] = newValue;
+                }
+            }
+        });
+
+        return oldState;
```

Comment:
Instead of mutating the passed in parameter (which lint rules will not allow), can you construct a new object?

Reply:
Yep, will update.